consensus/istanbul/backend/announce.go

```diff
@@ -589,130 +649,311 @@ func (sb *Backend) regossipAnnounce(msg *istanbul.Message, payload []byte) error
 	validatorConnSet, err := sb.retrieveValidatorConnSet()
 	if err != nil {
+		logger.Trace("Error in retrieving validator conn set", "err", err)
 		return err
 	}
 
-	announcesToRequest := make([]common.Address, 0)
-	for _, announceVersion := range announceVersions {
-		// Ignore this announceVersion entry if it's val address is not in the current validator connection set.
-		if !validatorConnSet[announceVersion.ValAddress] {
-			logger.Trace("Ignoring announceVersion since it's not in active/elected valset", "valAddress", announceVersion.ValAddress)
+	var validEntries []*vet.SignedAnnounceVersionEntry
+	validAddresses := make(map[common.Address]bool)
+	// Verify all entries are valid and remove duplicates
+	for _, signedAnnVersion := range signedAnnVersions {
+		err := signedAnnVersion.ValidateSignature()
+		if err != nil {
+			logger.Debug("Error validating signed announce version signature", "address", signedAnnVersion.Address, "err", err)
+			continue
+		}
+		if !validatorConnSet[signedAnnVersion.Address] {
+			logger.Debug("Found signed announce version from an address not in the validator conn set", "address", signedAnnVersion.Address)
+			continue
+		}
+		if _, ok := validAddresses[signedAnnVersion.Address]; ok {
+			logger.Debug("Found duplicate signed announce version in message", "address", signedAnnVersion.Address)
 			continue
 		}
+		validAddresses[signedAnnVersion.Address] = true
+		validEntries = append(validEntries, signedAnnVersion.Entry())
+	}
+	return sb.upsertSignedAnnounceVersionEntries(validEntries)
+}
 
-		sb.cachedAnnounceMsgsMu.RLock()
-		if cachedEntry, ok := sb.cachedAnnounceMsgs[announceVersion.ValAddress]; !ok || cachedEntry.MsgVersion < announceVersion.AnnounceMsgVersion {
-			announcesToRequest = append(announcesToRequest, announceVersion.ValAddress)
+func (sb *Backend) upsertSignedAnnounceVersionEntries(entries []*vet.SignedAnnounceVersionEntry) error {
+	logger := sb.logger.New("func", "upsertSignedAnnounceVersions")
+	newEntries, err := sb.signedAnnounceVersionTable.Upsert(entries)
+	if err != nil {
+		logger.Warn("Error in upserting entries", "err", err)
+	}
+
+	// If this node is a validator (checked later as a result of this call) and it receives a signed announce
+	// version from a remote validator that is newer than the remote validator's
+	// version in the val enode table, this node did not receive a direct announce
+	// and needs to announce its own enode to the remote validator.
+	sb.startGossipAnnounceTask()
+
+	// Only regossip entries that are new to this node and do not originate
+	// from an address that we have gossiped a signed announce version for
+	// within the last 5 minutes
+	var signedAnnVersionsToRegossip []*signedAnnounceVersion
+	sb.lastSignedAnnounceVersionsGossipedMu.Lock()
+	for _, entry := range newEntries {
+		lastGossipTime, ok := sb.lastSignedAnnounceVersionsGossiped[entry.Address]
+		if !ok || time.Since(lastGossipTime) >= 5*time.Minute {
```

Comment:
This is a key piece of codee and I feel its a little burried in here. We should extract this constant to the top of file, possibly even making it configurable, and make it more obvious that the DDoS protection is implemented here. Additionally, it would be more ideomatic if this was an `if ... { continue }` to decrease the indent level.